Documentation for Scrape Classes (#7299)
prometheus-operator/prometheus-operator: Documentation/developer/scrapeclass.md

---
weight: 254
toc: true
title: Scrape Classes
menu:
    docs:
        parent: developer
lead: ""
images: []
draft: false
description: null
---

<br>
<div class="alert alert-info" role="alert">

>     <i class="fa fa-exclamation-triangle"></i><b> Note:</b> Starting with v0.73.0, Scrape Class feature is added to Prometheus-Operator.

Reviewer: a bit redundant with the prerequisites section?

Author: Yup, will remove it.

</div>

## Prerequisites

Before you begin, ensure that you have:

* `Prometheus-Operator > v0.73.0`
* A running `Prometheus` or `PrometheusAgent` instance.

## Introduction

`ScrapeClass` is a feature that allows you to define common configuration settings to be applied across all scrape resources( **PodMonitor**, **ServiceMonitor**, **ScrapeConfig** and **Probe** ). This feature is similar to [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) in Kubernetes and is very useful when it comes to standardising configurations such as common relabelling rules, TLS certificates and authentication.

> One use-case is to configure authentication with TLS certificates when running `Prometheus` to scrape all the pods in an Istio mesh with [strict mTLS](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). Defining the TLS certificate paths in each `PodMonitor` and `ServiceMonitor` that scrapes these pods would be repetitive and error-prone. This problem is now solved by the `ScrapeClass` feature.

Reviewer: (nit) I would move this paragraph to the next section. WDYT?

Author: This use-case being in the introduction section gives an idea for why we need ScrapeClass. If we want to add multiple use cases as examples, I think having a separate section would be better but for just this one example, it is fine as it is.

Reviewer: :+1:

## Defining a ScrapeClass in Prometheus Resource

Below is an example of defining a scrape class in `Prometheus/PrometheusAgent` resource:

```yaml mdox-exec="cat example/user-guides/scrapeclass/scrapeclass-example-definition.yaml"
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
spec:
  scrapeClasses:
    - name: istio-mtls
      default: true
      tlsConfig:
        caFile: "/etc/istio-certs/root-cert.pem"
        certFile: "/etc/istio-certs/cert-chain.pem"
        keyFile: "/etc/istio-certs/key.pem"
        insecureSkipVerify: true

  # mount the certs from the istio sidecar (shown here for illustration purposes)
  volumeMounts:
    - name: istio-certs
      mountPath: "/etc/istio-certs/"
  volumes:
  - emptyDir:
      medium: Memory
    name: istio-certs
```

> An administrator can set the `default:true` so that the scrape applies to all scrape objects that don't configure an explicit scrape class. Only one scrape class can be set as default. If there are multiple default scrape classes, the operator will fail the reconciliation. The following log will be generated:

Reviewer: ```suggestion
An administrator can set the `default:true` so that the scrape applies to all scrape objects that don't configure an explicit scrape class. Only one scrape class can be set as default. If there are multiple default scrape classes, the operator will fail the reconciliation and the failure will be reported in the status conditions. 
```

apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: servicemonitor-example
spec:
  scrapeClass: istio-mtls
  endpoints:
    - port: http
      path: /metrics
```

If the monitor resource specifies a scrape class name that isn't defined in the `Prometheus/PrometheusAgent` object, then the operator will emit a Kubernetes event. Consider the following example where a `Prometheus` instance defines a scrapeClass, but a `ServiceMonitor` references a different one.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: my-prometheus
spec:
  replicas: 1
  scrapeClasses:
    - name: istio-mtls
      default: true
```

The `ServiceMonitor` references a scrapeClass named **istio**, which is not defined in the Prometheus object.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: example-service-monitor
spec:
  selector:
    matchLabels:
      app: example
  scrapeClass: istio
  endpoints:
    - port: http
      path: /metrics
```

As `ServiceMonitor` references a scrapeClass that does not exist in Prometheus, the following event is emitted:

> ```yaml

Reviewer: (nit) it's not a yaml code snippet :)

---
weight: 254
toc: true
title: Scrape Classes
menu:
    docs:
        parent: developer
lead: ""
images: []
draft: false
description: null
---

<br>
<div class="alert alert-info" role="alert">
    <i class="fa fa-exclamation-triangle"></i><b> Note:</b> Starting with v0.73.0, Scrape Class feature is added to Prometheus-Operator.
</div>

## Prerequisites

Before you begin, ensure that you have:

* `Prometheus-Operator > v0.73.0`

> * A running `Prometheus` instance.

Reviewer 2: Prometheus Agent as well?

## Introduction

`ScrapeClass` is a feature that allows you to define common configuration settings to be applied to all scrape resources( **PodMonitor**, **ServiceMonitor**, **ScrapeConfig** and **Probe** ) of that particular class. This feature is similar to [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/) in Kubernetes. This feature is very useful when it comes to standardizing configurations such as common relabeling rules, certificates, and other configurations.

> Traditionally, the scrape configurations were defined in the `Prometheus` config file itself. This process was time consuming when user wanted to apply the same configuration to many scrape targets. One such problem was when a user needed to configure secure scraping with TLS certificates when running `Prometheus` to scrape all the pods in an Istio mesh with strict `mTLS`, described [here](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). Defining the TLS certificate paths in each `PodMonitor` and `ServiceMonitor` that scrapes these pods would be repetitive and error-prone. This problem is now solved by `ScrapeClass` feature.

Reviewer 3: ```suggestion
Traditionally, the scrape configurations were defined in the scrape custom resources like ServiceMonitor  itself. This process was time consuming when a user wanted to apply the same configuration to many scrape targets. One use-case is when a user needed to configure secure scraping with TLS certificates when running `Prometheus` to scrape all the pods in an Istio mesh with strict `mTLS` described [here](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). Defining the TLS certificate paths in each `PodMonitor` and `ServiceMonitor` that scrapes these pods would be repetitive and error-prone. This problem is now solved by `ScrapeClass` feature.
```

```yaml mdox-exec="cat example/user-guides/scrapeclass/scrapeclass-example-servicemonitor.yaml"
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: servicemonitor-example
spec:
  scrapeClassName: istio-mtls
  endpoints:
    - port: http
      path: /metrics
```

If the monitor resource specifies a scrape class name that isn't defined in the `Prometheus/PrometheusAgent` object, then the scrape resource is rejected by the operator.

Similarly, we can select the scrape class for `PodMonitor` resource.

```yaml mdox-exec="cat example/user-guides/scrapeclass/scrapeclass-example-podmonitor.yaml"
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
spec:
  scrapeClassName: istio-mtls
  podMetricsEndpoints:
  - port: http
    path: /metrics
```

If a monitor resource includes a `tlsConfig` field, the Operator will apply a merge strategy to combine the `tlsConfig` fields from the monitor resource with those defined in the scrape class. The `tlsConfig` settings in the monitor resource will take precedence.

> 

Reviewer 3: ```suggestion
**Note: ** The configuration in scrapeClass will only be applied if the scrape resources haven't set fields defined in scrapeClass.
```

---
weight: 254
toc: true
title: Scrape Classes
menu:
    docs:
        parent: developer
lead: ""
images: []
draft: false
description: null
---

<br>
<div class="alert alert-info" role="alert">
    <i class="fa fa-exclamation-triangle"></i><b> Note:</b> Starting with v0.73.0, Scrape Class feature is added to Prometheus-Operator.
</div>

## Prerequisites

Before you begin, ensure that you have:

* `Prometheus-Operator > v0.73.0`
* A running `Prometheus` or `PrometheusAgent` instance.

## Introduction

> `ScrapeClass` is a feature that allows you to define common configuration settings to be applied across all scrape resources( **PodMonitor**, **ServiceMonitor**, **ScrapeConfig** and **Probe** ). This feature is similar to [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) in Kubernetes. This feature is very useful when it comes to standardising configurations such as common relabelling rules, certificates, and other configurations.

Reviewer: ```suggestion
`ScrapeClass` is a feature that allows you to define common configuration settings to be applied across all scrape resources( **PodMonitor**, **ServiceMonitor**, **ScrapeConfig** and **Probe** ). This feature is similar to [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) in Kubernetes. This feature is very useful when it comes to standardising configurations such as common relabelling rules, TLS certificates and authentication.
```

> Traditionally, the scrape configurations were defined in the scrape custom resources like ServiceMonitor itself. This process was time consuming when a user wanted to apply the same configuration to many scrape targets. One use-case is when a user needed to configure secure scraping with TLS certificates when running `Prometheus` to scrape all the pods in an Istio mesh with strict `mTLS` described [here](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). Defining the TLS certificate paths in each `PodMonitor` and `ServiceMonitor` that scrapes these pods would be repetitive and error-prone. This problem is now solved by `ScrapeClass` feature.

Reviewer: ```suggestion
One use-case is to configure authentication with TLS certificates when running `Prometheus` to scrape all the pods in an Istio mesh with [strict mTLS](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings). Defining the TLS certificate paths in each `PodMonitor` and `ServiceMonitor` that scrapes these pods would be repetitive and error-prone. This problem is now solved by the `ScrapeClass` feature.
```

## Defining a ScrapeClass in Prometheus Resource

Below is an example of defining a scrape class in `Prometheus/PrometheusAgent` resource:

```yaml mdox-exec="cat example/user-guides/scrapeclass/scrapeclass-example-definition.yaml"
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
spec:
  scrapeClasses:
    - name: istio-mtls
      default: true
      tlsConfig:
        caFile: "/etc/istio-certs/root-cert.pem"
        certFile: "/etc/istio-certs/cert-chain.pem"
        keyFile: "/etc/istio-certs/key.pem"
        insecureSkipVerify: true

  # mount the certs from the istio sidecar (shown here for illustration purposes)
  volumeMounts:
    - name: istio-certs
      mountPath: "/etc/istio-certs/"
  volumes:
  - emptyDir:
      medium: Memory
    name: istio-certs
```

An administrator can set the `default:true` so that the scrape applies to all scrape objects that don't configure an explicit scrape class. Only one scrape class can be set as default. If there are multiple default scrape classes, the operator will fail the reconciliation.

## Using the ScrapeClass in Monitor Resources

> Once the `ScrapeClass` is defined in the `Prometheus` resource, the `ScrapeClassName` field can be used in monitor resource to reference the particular `ScrapeClass`.

Reviewer: (nit) we use "scrape resources" to describe *Monitor resources. Let's try to stick to it.
```suggestion
Once the `ScrapeClass` is defined in the `Prometheus` resource, the `ScrapeClassName` field can be used in the scrape resource to reference the particular `ScrapeClass`.
```

```yaml mdox-exec="cat example/user-guides/scrapeclass/scrapeclass-example-servicemonitor.yaml"
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: servicemonitor-example
spec:
  scrapeClassName: istio-mtls
  endpoints:
    - port: http
      path: /metrics
```

> If the monitor resource specifies a scrape class name that isn't defined in the `Prometheus/PrometheusAgent` object, then the scrape resource is ignored by the operator.

Reviewer 2: Would be nice if we share a log example of the reconciling being failed, I don't remember if we're failing or not.

Author: I was able to get the log for multiple scrapeClass defined with `default:true` but could not get this one.

Reviewer: The operator will emit a Kubernetes event.

Author: I checked the function and it just returns an error

https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/resource_selector.go#L373C1-L385C2

Author: I tried but couldn't get the event. Maybe I made some mistakes. Here are my yaml files:

prometheus.yaml
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: my-prometheus
spec:
  replicas: 1
  scrapeClasses:
    - name: "istio-mtls"
    default: true
```

servciemonitor.yaml
```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: example-service-monitor
spec:
  selector:
    matchLabels:
      app: example
  scrapeClass: "istio"
  endpoints:
    - port: http
      path: /metrics
```

Reviewer: I'm seeing the event

> 0s          Warning   InvalidConfiguration   servicemonitor/prometheus-self                                                                           ServiceMonitor prometheus-self was rejec
ted due to invalid configuration: scrapeClass "foo" not found in Prometheus scrapeClasses

> If the monitor resource specifies a scrape class name that isn't defined in the `Prometheus/PrometheusAgent` object, then the scrape resource is rejected by the operator.

Reviewer: ```suggestion
If the monitor resource specifies a scrape class name that isn't defined in the `Prometheus/PrometheusAgent` object, then the scrape resource is ignored by the operator.
```
